### PR TITLE
Set default allocator id to device instead of unified for CUDA and HIP execution policies

### PR DESCRIPTION
**Pull request by bmhan12**
This PR:
- Makes device the default Umpire allocator id instead of unified for CUDA and HIP execution policies.
- Adjusts unit tests and examples to reflect device memory as the new default.
  - For tests and examples where device id for memory allocation did not work, unified id is used instead.

Follow-up work is required to revisit tests and examples that still require unified memory, and adjust the source implementation.  These are captured in https://github.com/LLNL/axom/issues/1339

### CODE REVIEW

File: src/axom/core/tests/core_execution_for_all.hpp
Comment (kennyweiss) [inline src/axom/core/tests/core_execution_for_all.hpp:1]: Suggestion: 
This file/test might be clearer / easier to understand with explicit allocators rather than using the (implicit) allocator from `getDefaultAllocatorID()`

(unless it's testing the implicit `getDefaultAllocator` functionality)

File: src/axom/mint/tests/mint_execution_cell_traversals.cpp
Comment (kennyweiss) [inline src/axom/mint/tests/mint_execution_cell_traversals.cpp:1]: It looks like the `mint_execution_cell_traversals` require unified memory. 
Would it make sense to add a utility function at the top of the file to reduce the duplication?

**Edit:**  And similarly in the other mint files

File: src/axom/primal/tests/primal_boundingbox.cpp [hunk 1]
```diff
@@ -40,7 +40,10 @@ void check_bb_policy()
       box[i].isValid();
     });
 
-  EXPECT_EQ(box[0], BoundingBoxType(PointType(0.0), PointType(10.0)));
+  BoundingBoxType box_host;
```
Comment (kennyweiss) [inline src/axom/primal/tests/primal_boundingbox.cpp:43]: Minor: Would this code be simpler if  `box` used a `StackArray`  instead of allocating and deallocating one bounding box?

File: src/axom/primal/tests/primal_clip.cpp [hunk 2]
```diff
@@ -329,6 +329,12 @@ void unit_check_poly_clip()
   //      |
   // In addition, vertices 0 and 3 should be marked as clipped.
 
+  const int current_allocator = axom::getDefaultAllocatorID();
```
Comment (kennyweiss) [inline src/axom/primal/tests/primal_clip.cpp:332]: Similarly here: 
Do we have a better pattern for allocating/deallocating moving a single object between host and device, e.g. using a `StackArray` ?

(Not required for this PR, but if we have a cleaner way of doing this, we should prefer to use it in the unit tests)

File: src/axom/primal/tests/primal_clip.cpp [hunk 3]
```diff
@@ -434,46 +439,63 @@ void check_hex_tet_clip(double EPS)
...
+                               PointType {1, 0, 1});
+
+      hex[0] = HexahedronType(PointType {0, 0, 0},
+                              PointType {1, 0, 0},
+                              PointType {1, 1, 0},
+                              PointType {0, 1, 0},
+                              PointType {0, 0, 1},
+                              PointType {1, 0, 1},
+                              PointType {1, 1, 1},
+                              PointType {0, 1, 1});
```
Comment (kennyweiss) [inline src/axom/primal/tests/primal_clip.cpp:462]: Does this change require `setDefaultAllocatorID()`  to have been set above? Or does it work because its constructed within the appropriate `ExecPolicy` ?

If it's the former, is there a way to make this work without setting the default allocator?

File: src/axom/primal/tests/primal_clip.cpp [hunk 4]
```diff
@@ -503,44 +525,60 @@ void check_oct_tet_clip(double EPS)
...
-  oct[0] = OctahedronType(PointType {1, 0, 0},
-                          PointType {1, 1, 0},
-                          PointType {0, 1, 0},
-                          PointType {0, 1, 1},
-                          PointType {0, 0, 1},
-                          PointType {1, 0, 1});
+  // Shapes on host
+  TetrahedronType tet_host;
+  OctahedronType oct_host;
+  PolyhedronType res_host;
```
Comment (kennyweiss) [inline src/axom/primal/tests/primal_clip.cpp:531]: Likely outside the scope of this PR, but we should probably have a standardized convention for labeling variables as host, device and views and use them consistently within axom and its tests and examples.

Perhaps we should create a separate issue to track this?

File: src/axom/primal/tests/primal_polyhedron.cpp [hunk 5]
```diff
@@ -290,43 +290,48 @@ void check_volume()
...
-  polys[0].addNeighbors(0, {1, 4, 3});
-  polys[0].addNeighbors(1, {5, 0, 2});
-  polys[0].addNeighbors(2, {3, 6, 1});
-  polys[0].addNeighbors(3, {7, 2, 0});
-  polys[0].addNeighbors(4, {5, 7, 0});
-  polys[0].addNeighbors(5, {1, 6, 4});
-  polys[0].addNeighbors(6, {2, 7, 5});
-  polys[0].addNeighbors(7, {4, 6, 3});
+  // Volume on host (for CUDA or HIP)
+  double res_host;
```
Comment (kennyweiss) [inline src/axom/primal/tests/primal_polyhedron.cpp:304]: Minor (clarity): consider renaming these variables from `res` to `volume`

File: src/axom/quest/IntersectionShaper.hpp [hunk 6]
```diff
@@ -162,7 +162,10 @@ class GridFunctionView
     m_hostData = hostPtr;
     m_numElements = nElem;
     m_needResult = _needResult;
-    int execSpaceAllocatorID = axom::execution_space<ExecSpace>::allocatorID();
+    constexpr bool on_device = axom::execution_space<ExecSpace>::onDevice();
+    int execSpaceAllocatorID = on_device
+      ? axom::getUmpireResourceAllocatorID(umpire::resource::Unified)
+      : axom::execution_space<ExecSpace>::allocatorID();
```
Comment (kennyweiss) [inline src/axom/quest/IntersectionShaper.hpp:168]: Suggestion: Consider a (private) utility/helper function for this

File: src/axom/spin/tests/spin_implicit_grid.cpp
Comment (kennyweiss) [inline src/axom/spin/tests/spin_implicit_grid.cpp:1]: Minor: This file seems to have a different convention than other files in axom

**Review by kennyweiss — APPROVED** [review]
Thanks @bmhan12 !

File: src/axom/primal/tests/primal_clip.cpp (hunk 3 above)
Comment (bmhan12) [reply src/axom/primal/tests/primal_clip.cpp:462]: > Does this change require setDefaultAllocatorID() to have been set above? Or does it work because its constructed within the appropriate ExecPolicy ?

The latter, and it works because the hexahedron constructor is host-device decorated.

File: src/axom/primal/tests/primal_clip.cpp (hunk 2 above)
Comment (bmhan12) [reply src/axom/primal/tests/primal_clip.cpp:332]: > Do we have a better pattern for allocating/deallocating moving a single object between host and device, e.g. using a StackArray ?

Not that I am aware of. The best I have found is either using `axom::allocate` & `axom::copy`, or using `axom::Array` constructors to move between host and device.

File: src/axom/primal/tests/primal_clip.cpp (hunk 4 above)
Comment (bmhan12) [reply src/axom/primal/tests/primal_clip.cpp:531]: Issue #1333 opened to track this issue.

File: src/axom/core/tests/core_array_for_all.hpp [hunk 7]
```diff
@@ -1204,6 +1204,15 @@ AXOM_TYPED_TEST(core_array_for_all, device_insert)
     typename TestFixture::template DynamicTArray<DynamicArray>;
 
   int kernelAllocID = axom::execution_space<ExecSpace>::allocatorID();
+#if defined(AXOM_USE_GPU) && defined(AXOM_USE_UMPIRE)
+  // Use unified memory for frequent movement between device operations
+  // and value checking on host
+  if(axom::execution_space<ExecSpace>::onDevice())
```
Comment (rhornung67) [inline src/axom/core/tests/core_array_for_all.hpp:1210]: Would it be helpful to remove the exec space check to make sure everything keep working for the "degenerate case" where data is only used on the host?

File: src/axom/mint/tests/mint_execution_cell_traversals.cpp [hunk 8]
```diff
@@ -30,6 +30,22 @@ namespace mint
 //------------------------------------------------------------------------------
 namespace
 {
+#if defined(AXOM_USE_RAJA) && defined(AXOM_USE_UMPIRE) &&   \
```
Comment (rhornung67) [inline src/axom/mint/tests/mint_execution_cell_traversals.cpp:33]: It looks like all this code is the same as the code in the files for mint tests of face traversals and node traversals. It seems like it would save a lot of maintenance if it were centralized in one place and reused. @bmhan12 if you agree, this could be done in a subsequent PR.

**Review by rhornung67 — APPROVED** [review]
A couple of comments about catching corner cases and code consolidation that could be done in another PR.

File: src/axom/mint/tests/mint_execution_cell_traversals.cpp (hunk 8 above)
Comment (bmhan12) [reply src/axom/mint/tests/mint_execution_cell_traversals.cpp:33]: Added this as a task part of #1339
